google/cloud/internal/external_account_token_source_aws.cc

```diff
@@ -80,9 +85,20 @@ StatusOr<internal::SubjectToken> Source(
   if (!region) return std::move(region).status();
   auto secrets = FetchSecrets(info, *token, cf, opts, ec);
   if (!secrets) return std::move(secrets).status();
-  auto now = std::chrono::system_clock::now();
-  auto subject = ComputeSubjectToken(info, *region, *secrets, now, audience);
-  return internal::SubjectToken{subject.dump()};
+  auto const now = std::chrono::system_clock::now();
+  auto const subject =
+      ComputeSubjectToken(info, *region, *secrets, now, audience).dump();
+  // We need to URL-encode the string.
+  using CharMapping = std::pair<absl::string_view, absl::string_view>;
+  auto const mappings = std::array<CharMapping, 25>{{
+      {" ", "%20"}, {"\"", "%22"}, {"#", "%23"},  {"$", "%24"}, {"%", "%25"},
+      {"&", "%26"}, {"+", "%2B"},  {",", "%2C"},  {"/", "%2F"}, {":", "%3A"},
+      {";", "%3B"}, {"<", "%3C"},  {"=", "%3D"},  {">", "%3E"}, {"?", "%3F"},
+      {"@", "%40"}, {"[", "%5B"},  {"\\", "%5C"}, {"]", "%5D"}, {"^", "%5E"},
+      {"`", "%60"}, {"{", "%7B"},  {"|", "%7C"},  {"}", "%7D"}, {"\177", "%7F"},
+  }};
+
+  return internal::SubjectToken{absl::StrReplaceAll(subject, mappings)};
```

Comment:
Possible followup: Perhaps we should make an `internal/url_encoder.cc`.

Reply:
Ack.